### PR TITLE
Golang native filemonitor

### PR DESCRIPTION
**Pull request by metcalf**
Switches filemonitor to be written in purely Golang (with C hooks where required) rather than shelling out to binaries compiled to C++. As evidence that this is an improvement, I was able to cut the `time.Sleep` in the integration test in half and successfully reload multiple services during that test. The `time.Sleep` isn't even necessary on OS X anymore, just with inotify on TravisCI.

In the process I had to make several other changes:
* Merged the `restarter` logic into the filemonitor. FSEvents already provides for debouncing so we were previously double-debouncing.
* Refactored the filemonitor interface to be testable and reusable rather than having lots of global state.
* Added tests for all of the filemonitors. Note that since the monitors compile conditionally, the fsevents-based monitor can't run on TravisCI and the inotify-based monitor can't run in local OS X development.
* Fixed a bug in monitoring for FSEvents where the EventID was not maintained between restarts, potentially dropping events.

TODO:
* [x] Introduce vendoring of dependencies. The fsevents and fsnotify libraries do not have any API compatibility guarantees right now. Pending merge and rebase on https://github.com/burke/zeus/pull/553
* [x] Actually use the file-listener when the env var is set.
* [x] Find a way to integration test the file-listener (at least locally, one-off).
* [x] Don't double-debounce fsevents.
* [x] Delete unused makefile code

### CODE REVIEW

File: go/filemonitor/filelistener.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	case <-f.stop:
+		f.Unlock()
+		return nil // Already stopped
+	default:
+		close(f.stop)
+	}
+
+	var firstErr error
+	if firstErr = f.netListener.Close(); firstErr != nil {
+		slog.Trace("Error closing file listener: %v", firstErr)
```
Comment (ptarjan) [inline go/filemonitor/filelistener.go]: where do these slogs do during normal operation?

Comment (metcalf) [inline go/filemonitor/filelistener.go]: If you pass `--log` to Zeus, it logs them to whatever file you provide. In our setup, you can read these trace logs in `dev/zeus.log`... They're pretty helpful for debugging now that I found them!

File: go/filemonitor/filemonitor_darwin.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
+// +build darwin
+
+package filemonitor
+
+import "github.com/fsnotify/fsevents"
+
+const flagsWorthReloadingFor = fsevents.ItemRemoved | fsevents.ItemModified | fsevents.ItemRenamed
```
Comment (ptarjan) [inline go/filemonitor/filemonitor_darwin.go]: weird how there isn't an added https://github.com/fsnotify/fsevents/blob/master/fsevents.go#L91

**Comment by ptarjan** [comment]
lgtm

My go isn't stellar but it looked all reasonable. The locking was a little hairy but seemed right.

File: go/filemonitor/filemonitor_darwin.go (hunk 2 above)
Comment (metcalf) [inline go/filemonitor/filemonitor_darwin.go]: There is an `ItemCreated`. I copied this list from here: https://github.com/burke/zeus/blob/ff7c2559dc2bc617a4ea2603f04aacd277f7a5a3/ext/fsevents/main.m#L11-L13

There aren't a lot of people talking about the subtleties of FSEvents but a couple places seem to indicate that you'll get an `kFSEventStreamEventFlagItemCreated` followed by an `kFSEventStreamEventFlagItemModified` if you create a file:
http://stackoverflow.com/questions/15194662/file-system-watcher-for-mac-os-x
http://stackoverflow.com/questions/11123007/check-which-files-have-changed-in-macfsevents-in-perl

**Pull request merged**